mirage-router/unikernel.ml

```diff
@@ -0,0 +1,148 @@
+(* setup is as follows:
+   the stack `s` is used for the openvpn tunnel
+   on the private network interface we are a gateway (for our private network):
+   - each packet received there that is addressed to us (at the ethernet layer) is forwarded via the tunnel [fragmentation may apply]
+   - a packet read from the tunnel is forwarded onto the private network interface [reassembly may apply]
+
+   --> the 'cb' below is called for each IPv4 frame that we received over the tunnel:
+    - figure out depending on "dst" whether it needs to be forwarded via private network interface
+    -> we can use Static_ipv4.write (with ~src) for sending out ipv4 packets
+   --> another task needs to listen on the private network needs to send IPv4 packets from the private network over the tunnel
+    -> we need a custom "listen" on ipv4 (arp as usual) which does not discard packets which are not addressed to us
+*)
+open Lwt.Infix
+
+module Main (R : Mirage_random.S) (M : Mirage_clock.MCLOCK) (P : Mirage_clock.PCLOCK) (T : Mirage_time.S) (S : Mirage_stack.V4)
+    (N : Mirage_net.S) (E : Mirage_protocols.ETHERNET) (A : Mirage_protocols.ARP) (I : Mirage_protocols.IPV4) (FS: Mirage_kv.RO) = struct
+
+  module O = Openvpn_mirage.Make(R)(M)(P)(T)(S)
+
+  let read_config data =
+    FS.get data (Mirage_kv.Key.v "openvpn.config") >|= function
+    | Error e -> Rresult.R.error_to_msg ~pp_error:FS.pp_error (Error e)
+    | Ok data ->
+      let string_of_file _ = Error (`Msg "not supported") in
+      Openvpn.Config.parse_client ~string_of_file data
+
+  let local_network ip =
+    let net, my_ip = Key_gen.private_ipv4 () in
+    if Ipaddr.V4.compare my_ip ip = 0 then begin
+      Logs.warn (fun m -> m "a packet directed to us (ignoring)");
+      false
+    end else
+      Ipaddr.V4.Prefix.mem ip net
+
+  type t = {
+    ovpn : O.t ;
+    mutable ovpn_fragments : Fragments.Cache.t ;
+    private_ip : I.t ;
+    mutable private_fragments : Fragments.Cache.t ;
+  }
+
+  let private_recv net eth arp t =
+    let our_listen =
+      E.input
+        ~arpv4:(A.input arp)
+        ~ipv4:(fun _ -> Logs.warn (fun m -> m "ignoring IPv4"); Lwt.return_unit)
+        ~ipv6:(fun _ -> Logs.warn (fun m -> m "ignoring IPv6"); Lwt.return_unit)
+        eth
+    and forward_packet ip_hdr payload =
+      let c, pkt = Fragments.process t.private_fragments (M.elapsed_ns ()) ip_hdr payload in
+      t.private_fragments <- c;
+      Logs.info (fun m -> m "%B forwarding packet %a"
+                    (match pkt with None -> false | Some _ -> true)
+                    Ipv4_packet.pp ip_hdr);
+      match pkt with
+      | None -> Lwt.return_unit
+      | Some (hdr, pay) ->
+        (* we need to check mtu and potentially fragment *)
```

Comment:
I plan to do this today. `forward_packet` is called when a packet on the private network interface is received which should be sent over the tunnel. some lines above, we push it through the reassembly cache for the private network -- this means the full packet is now recovered (if you send a huge ping, this is completely reassembled at this point). Now, the OpenVPN tunnel likely has a smaller MTU (there's mac and padding overhead) than the private network (usually 1500 bytes) one, thus the payload needs to be cut into pieces.

An underlying issue is that some IP packets should not be fragmented, but instead an error should be sent to the sender (this is crucial for path MTU discovery, and needs to be handled here).

For the OpenVPN tunnel, there's in `Openvpn_mirage.Make()` a `  val mtu : t -> int` to be used here.

Reply:
after thinking a bit harder, since this is a gateway, to behave nicely in a network, the TTL field should be reduced by one, and in case the TTL turns 0, another ICMP error should be sent to the sender (instead of forwarding the packet) -- I'll implement this as well. maybe one day we should put such code in a different library and reuse it across implementations (I suspect the NAT library does something along the lines already).

Reply:
done in ed5d2b6